File: ruby/5.rb
# 2520 is the smallest number that can be divided by each of the numbers from 1
# to 10 without any remainder.

# What is the smallest positive number that is evenly divisible by all of the
# numbers from 1 to 20?

def IsDivisibleUpTo(x)
  # checks if it is divisible up to max
  max = 20
  for i in 2..max
    if x % i != 0
      return false
    end
  end
  return true
end

def brutishForcish(tldr)
  ans = []
  for i in 2..tldr
    if IsDivisibleUpTo(i)
      ans << i
    end
    if ans.length > 0
      return i
    end
  end
end

puts brutishForcish(252000000)

# prolly better to build an array that is only multiples primes, 
# then check only those values against the non-prime 1-20

File: rust/src/5.rs
// 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

// What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

fn main() {
    let result = lcm_of_all_below(10);
    println!("smallest positive number evenly divisible by all numbers from 1 to 20: {}", result);
}

fn lcm_of_all_below(max: u32) -> u32 {
    let min = max / 2;

    return acc_min_factors(factor_map.init(), max, min);
}

fn acc_min_factors(factor_map: FactorMap, next: u32, min: u32) {
    if next <= min {
        return factor_map.total();
    }

    let next_fm = factor_map.from(next);

    let factor_map = factor_map.merge(next_fm);

    return acc_min_factors(factor_map, next - 1, min);
}
